src/lib/transition_router/initial_validator.ml

```diff
@@ -233,142 +232,109 @@ module Duplicate_block_detector = struct
           [%log error] ~metadata msg )
 end
 
-let run ~logger ~trust_system ~verifier ~transition_reader
-    ~valid_transition_writer ~initialization_finish_signal ~precomputed_values =
-  let genesis_state_hash =
-    (Precomputed_values.genesis_state_hashes precomputed_values).state_hash
-  in
-  let genesis_constants =
-    Precomputed_values.genesis_constants precomputed_values
-  in
-  let rejected_blocks_logger =
-    Logger.create ~id:Logger.Logger_id.rejected_blocks ()
-  in
-  let open Deferred.Let_syntax in
-  let duplicate_checker = Duplicate_block_detector.create () in
-  O1trace.background_thread "initially_validate_blocks" (fun () ->
-      Reader.iter transition_reader
-        ~f:(fun
-             ( `Transition transition_env
-             , `Time_received time_received
-             , `Valid_cb valid_cb )
-           ->
-          let state_hash =
-            ( Envelope.Incoming.data transition_env
-            |> Mina_block.header |> Header.protocol_state
-            |> Protocol_state.hashes )
-              .state_hash
+let validate ~logger ~trust_system ~verifier ~initialization_finish_signal
+    ~precomputed_values =
+  stage (fun ~transition_env ~time_received ~valid_cb ->
```

Comment:
What's `stage`/`unstage`? Not familiar with these helper functions.

Reply:
They're built into Jane Street's `core` stdlib.

Reply:
For some reason there were no docs for it, but I've found some for `Staged.t` type.

https://ocaml.org/p/base/v0.17.0/doc/Base/Staged/index.html#type-t

I wonder, is there a benefit in using this special type instead of plain function?

Reply:
Using this special type stops you from passing all of the arguments at once. As you pointed out above, we really want to do some one-time setup, stage that environment, and then call the function using that environment multiple times.

Reply:
Classic example
```ocaml
let log_total () =
  let running_total = ref 0 in
  stage (fun increment ->
    running_total := !running_total + increment;
    Format.eprintf "Total is now %i" !running_total )

let () =
  let log_total = unstage @@ log_total () in
  log_total 2;
  log_total 3;
  log_total 5
```

In general though, this is good practice for any function that returns another function, to guide your users into not re-initialising every time (e.g. `log_total () 2` in the above example).